osu.Game/Rulesets/Edit/Checks/CheckDelayedHitsounds.cs

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using osu.Framework.Audio.Track;
+using osu.Game.Audio;
+using osu.Game.Extensions;
+using osu.Game.Rulesets.Edit.Checks.Components;
+
+namespace osu.Game.Rulesets.Edit.Checks
+{
+    public class CheckDelayedHitsounds : ICheck
+    {
+        /// <summary>
+        /// Threshold at which point the sample is considered silent.
+        /// </summary>
+        private const float silence_threshold = 0.001f;
+
+        private const float falloff_factor = 0.95f;
+        private const int delay_threshold = 5;
+        private const int delay_threshold_negligible = 1;
+
+        private readonly string[] audioExtensions = { "mp3", "ogg", "wav" };
+        private readonly string[] bankNames = { HitSampleInfo.BANK_NORMAL, HitSampleInfo.BANK_SOFT, HitSampleInfo.BANK_DRUM };
+        private readonly string[] sampleSets = { HitSampleInfo.HIT_NORMAL, HitSampleInfo.HIT_WHISTLE, HitSampleInfo.HIT_FINISH, HitSampleInfo.HIT_CLAP };
+
+        public CheckMetadata Metadata => new CheckMetadata(CheckCategory.Audio, "Delayed hit sounds.");
+
+        public IEnumerable<IssueTemplate> PossibleTemplates => new IssueTemplate[]
+        {
+            new IssueTemplateConsequentDelay(this),
+            new IssueTemplateDelay(this),
+            new IssueTemplateDelayNoSilence(this),
+            new IssuTemplateMinorDelay(this),
+            new IssuTemplateMinorDelayNoSilence(this),
+        };
+
+        private float getAverageAmplitude(Waveform.Point point) => (point.AmplitudeLeft + point.AmplitudeRight) / 2;
+
+        public IEnumerable<Issue> Run(BeatmapVerifierContext context)
+        {
+            var beatmapSet = context.Beatmap.BeatmapInfo.BeatmapSet;
+
+            if (beatmapSet == null)
+                yield break;
+
+            foreach (var file in beatmapSet.Files)
+            {
+                using (Stream? stream = context.WorkingBeatmap.GetStream(file.File.GetStoragePath()))
+                {
+                    if (stream == null)
+                        continue;
+
+                    if (!hasAudioExtension(file.Filename))
+                        continue;
+
+                    if (!isHitSound(file.Filename))
+                        continue;
+
+                    Waveform waveform = new Waveform(stream);
+
+                    var points = waveform.GetPoints();
+
+                    // Skip muted samples
+                    if (points.Length == 0 || points.Sum(getAverageAmplitude) <= silence_threshold)
+                        continue;
+
+                    float maxAmplitude = points.Select(getAverageAmplitude).Max();
```

Comment:
This isn't exactly ported from MV is it? MV seems to be using absolute value of sum over both channels rather than average: https://github.com/Naxesss/MapsetChecks/blob/8efbf76d17aa495732b2de0b75e33132169cecd4/Checks/AllModes/General/Audio/CheckHitSoundDelay.cs#L133

Reply:
Looking at MV source, it internally [uses `BASS_ChannelGetLevel()`](https://github.com/Naxesss/MapsetVerifierFramework/blob/60667b2ec73100265bd1687697cf4121e6d79a1a/objects/resources/AudioBASS.cs#L117C10-L117C10), which [as per docs should not return anything negative unless error](https://www.un4seen.com/doc/#bass/BASS_ChannelGetLevel.html):

> The level ranges linearly from 0 (silent) to 32768 (max). 0 will be returned when a channel is stalled.

So the "absolute value" part seems bogus, but the "sum" part still seems relevant. @Naxesss are you able to provide some context here? There isn't any in MV blame.

Reply:
Guessing things here. Is it because if a sound feels louder when played on both channels, so getting the sum of channel amplitudes guarantees that we would get the loudest peak possible?

Reply:
Looking into this again, this appears to be even more complicated than I thought, because osu!framework internally uses [`BASS_ChannelGetData()`](https://www.un4seen.com/doc/#bass/BASS_ChannelGetData.html), which can apparently return raw sample values (including in the negatives) unless one of the flags is specified, and I don't think any flags are being specified.

Needs a proper start-to-end investigation, probably with sample test cases provided. And this implementation is likely going to behave differently to mapsetverifier. I will not be doing this as this is a low-priority item; @ItsShamed if you wish to see this change merged you're going to have to do the legwork on this.

Reply:
I've looked into this, and upon looking at framework code, it kinda does what MV was doing in the attached snippet. 

https://github.com/ppy/osu-framework/blob/9ca2a05e3246d525852b395443af3782796f249b/osu.Framework/Audio/Track/Waveform.cs#L133-L15

So the points in the waveform are guaranteed to not return negative values at least.

At the very high level, this is the exact same logic, so in theory there should be not much of a discrepancy between MV and lazer. 
I tested it myself by adding the test samples into an already ranked beatmap (to not bloat checks) and passed it against MV and lazer: they both yield the same checks (apart some ms discrepancy)

I will retest that and provide screenshots later when I'll get home because screenshots are more concrete than words.

Reply:
For what it's worth, the check MV uses for delayed hitsounds is not ideal, as it often produces issues which could be considered false-positives for echo-y sound types, for example bells and chimes, where "the peak" is ambiguous. I've also noticed that the threshold/decay I use falls apart somewhat for quiet sound files, typically producing false-positives there too.

So if you have some approach in mind that might work better, then don't feel obligated to maintain parity with MV here.